src/utils.js

```diff
@@ -31,3 +31,53 @@ export function createColorGenerator(min_value, max_value, color) {
   }
   return colorGenerator;
 }
+
+// Create 'summary' dictionary: Record<string, {name: string; value: number}>
+function createSummaryDictionary(summariesOrDetails) {
+  return summariesOrDetails.reduce((summary, item) => {
+    if (summary[item.name] === undefined) {
+      summary[item.name] = {
+        name: item.name,
+        value: item.value,
+      };
+    } else {
+      summary[item.name].value += item.value;
+    }
+    return summary;
+  }, {});
+}
+
+// Sort summary dictionary
+function sortSummaryDictionary(summaryDictionary) {
+  return Object.values(summaryDictionary).sort((a, b) => {
+    return b.value - a.value;
+  });
+}
+
+export function getYearSummary(data, date) {
+  /**
+   * Filtering the 'data' based on the year of date,
+   * extracting all the summaries from the data into a single dimensional array and
+   */
+  const summaries = data
+    .filter(
+      (e) => new Date(e.date).getFullYear() === new Date(date).getFullYear()
+    )
+    .flatMap((e) => e.summary);
+  const summaryDictionary = createSummaryDictionary(summaries);
+  return sortSummaryDictionary(summaryDictionary);
+}
+
+// Calculate daily summary if that was not provided
+export function parseData(data) {
+  if (Array.isArray(data)) {
+    if (data[0].summary === null || data[0].summary === undefined) {
+      data.forEach((d) => {
```

Comment:
refactor to use map, and then update this.props.data using the fucntion output

Reply:
i mean `.map(...)`

Reply:
how the map function output will be used to update `data`, for that we would have to use another loop to iterate over data and update `summary` one my one on all `data elements`.

Reply:
simpler solution may be possible, think harder